toolbox/relnotes/main.go

```diff
@@ -0,0 +1,606 @@
+// Copyright 2017 The Kubernetes Authors All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/go-github/github"
+	u "k8s.io/release/toolbox/util"
+)
+
+var (
+	// Flags
+	// TODO: golang flags and parameters syntex
+	branch = flag.String("branch", "", "Specify a branch other than the current one")
+	full   = flag.Bool("full", false, "Force 'full' release format to show all sections of release notes. "+
+		"(This is the *default* for new branch X.Y.0 notes)")
+	githubToken   = flag.String("github-token", "", "Must be specified, or set the GITHUB_TOKEN environment variable")
+	htmlFileName  = flag.String("html-file", "", "Produce a html version of the notes")
+	htmlizeMD     = flag.Bool("htmlize-md", false, "Output markdown with html for PRs and contributors (for use in CHANGELOG.md)")
+	mdFileName    = flag.String("markdown-file", "", "Specify an alt file to use to store notes")
+	owner         = flag.String("owner", "kubernetes", "Github owner or organization")
+	preview       = flag.Bool("preview", false, "Report additional branch statistics (used for reporting outside of releases)")
+	releaseBucket = flag.String("release-bucket", "kubernetes-release", "Specify gs bucket to point to in generated notes (informational only)")
+	releaseTars   = flag.String("release-tars", "", "Directory of tars to sha256 sum for display")
+	repo          = flag.String("repo", "kubernetes", "Github repository")
+
+	// Global
+	branchHead = ""
+)
+
+func main() {
```

Comment:
Put all steps directly in `main()` is not a good idea, `main()` usually only call major high-level functions. You may want to make your program "Hierarchical".

Reply:
Currently `main()` contains mostly high-level logic in my opinion, but all of the release-note-related logic (functions) are in one file: main.go. Maybe I should put all the other functions except `main()` in some other files?

Reply:
Rearranged `main()`